tests/migrations/test_operations.py

```diff
@@ -241,6 +241,274 @@ def test_create_model_with_db_level_fk(self):
+
+        pony = Pony.objects.create()
+        kwargs = {fk_fieldname: pony}
+        Rider.objects.create(**kwargs)
+
+        with self.assertRaises(IntegrityError):
+            pony.delete()
+
+        Rider.objects.all().delete()
+        Pony.objects.all().delete()
+        with connection.schema_editor() as editor:
+            operation.database_backwards(app_label, editor, new_state, project_state)
+
+    def change_to_db_set_null_test(
+        self, app_label, fk_parent_model, fk_child_model, fk_fieldname, project_state
+    ):
+        operation = migrations.AlterField(
+            fk_child_model,
+            fk_fieldname,
+            models.ForeignKey(
+                f"{app_label}.{fk_parent_model}",
+                null=True,
+                on_delete=models.DB_SET_NULL,
+            ),
+        )
+        new_state = project_state.clone()
+        operation.state_forwards(app_label, new_state)
+
+        with connection.schema_editor() as editor:
+            operation.database_forwards(app_label, editor, project_state, new_state)
+
+        Rider = new_state.apps.get_model(app_label, fk_child_model)
+        Pony = new_state.apps.get_model(app_label, fk_parent_model)
+
+        pony = Pony.objects.create()
+        kwargs = {fk_fieldname: pony}
+        rider = Rider.objects.create(**kwargs)
+
+        with self.assertNumQueries(1):
+            pony.delete()
+
+        rider.refresh_from_db()
+        self.assertEqual(getattr(rider, fk_fieldname), None)
+
+        Rider.objects.all().delete()
+        Pony.objects.all().delete()
+        with connection.schema_editor() as editor:
+            operation.database_backwards(app_label, editor, new_state, project_state)
+
+    def change_to_db_set_default_test(
+        self, app_label, fk_parent_model, fk_child_model, fk_fieldname, project_state
+    ):
+        operation = migrations.AlterField(
+            fk_child_model,
+            fk_fieldname,
+            models.ForeignKey(
+                f"{app_label}.{fk_parent_model}",
+                db_default=1,
+                on_delete=models.DB_SET_DEFAULT,
+            ),
+        )
+        new_state = project_state.clone()
+        operation.state_forwards(app_label, new_state)
+
+        with connection.schema_editor() as editor:
+            operation.database_forwards(app_label, editor, project_state, new_state)
+
+        Rider = new_state.apps.get_model(app_label, fk_child_model)
+        Pony = new_state.apps.get_model(app_label, fk_parent_model)
+
+        pony1 = Pony.objects.create(id=1)
+        pony = Pony.objects.create()
+        kwargs = {fk_fieldname: pony}
+        rider = Rider.objects.create(**kwargs)
+
+        # Extra begin and commit queries
+        with self.assertNumQueries(3):
+            pony.delete()
+
+        rider.refresh_from_db()
+        pony1.refresh_from_db()
+        self.assertEqual(getattr(rider, fk_fieldname), pony1)
```

Comment:
Sounds interesting, will add that.

Reply:
Test : 
```
    @skipUnlessDBFeature("has_on_delete_db_default")
    def test_foreign_key_db_default(self):
        # Default parent exists
        default_parent = Foo.objects.create(pk=1)
        parent = Foo.objects.create(pk=2)
        child1 = DBDefaultsFK.objects.create(language_code=parent)
        with self.assertNumQueries(1):
            parent.delete()
        child1.refresh_from_db()
        self.assertEqual(child1.language_code, default_parent)
        child1.delete()
        default_parent.delete()

        # default parent dne
        parent = Foo.objects.create(pk=2)
        child1 = DBDefaultsFK.objects.create(language_code=parent)
        parent.delete()
        DBDefaultsFK.objects.all()
```

```
ERROR: test_foreign_key_db_default (delete.tests.DatabaseLevelOnDeleteTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\utils.py", line 103, in _execute
    return self.cursor.execute(sql)
psycopg2.errors.ForeignKeyViolation: insert or update on table "delete_dbdefaultsfk" violates foreign key constraint "delete_dbdefaultsfk_language_code_id_04474ba9_fk_delete_foo_id"
DETAIL:  Key (language_code_id)=(1) is not present in table "delete_foo".


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 294, in _setup_and_call
    self._post_teardown()
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 1112, in _post_teardown
    self._fixture_teardown()
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 1334, in _fixture_teardown
    connections[db_name].check_constraints()
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\postgresql\base.py", line 383, in check_constraints
    cursor.execute("SET CONSTRAINTS ALL IMMEDIATE")
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\utils.py", line 79, in execute
    return self._execute_with_wrappers(
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\utils.py", line 103, in _execute
    return self.cursor.execute(sql)
django.db.utils.IntegrityError: insert or update on table "delete_dbdefaultsfk" violates foreign key constraint "delete_dbdefaultsfk_language_code_id_04474ba9_fk_delete_foo_id"
DETAIL:  Key (language_code_id)=(1) is not present in table "delete_foo".
```

Reply:
Let me know if that is expected.

Reply:
Yeah, looks correct to me.

Reply:
Getting some weird results here.
**The test : **
```
    @skipUnlessDBFeature("has_on_delete_db_default")
    def test_foreign_key_db_default(self):
        # Default parent exists
        default_parent = Foo.objects.create(pk=1)
        parent = Foo.objects.create(pk=2)
        child1 = DBDefaultsFK.objects.create(language_code=parent)
        with self.assertNumQueries(1):
            parent.delete()
        child1.refresh_from_db()
        self.assertEqual(child1.language_code, default_parent)
        child1.delete()
        default_parent.delete()

        # default parent dne
        with self.assertRaises(IntegrityError):
            parent = Foo.objects.create(pk=2)
            child1 = DBDefaultsFK.objects.create(language_code=parent)
            parent.delete()
            DBDefaultsFK.objects.all()
```

**The error(or whatever)**
```
ERROR: test_foreign_key_db_default (delete.tests.DatabaseLevelOnDeleteTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 294, in _setup_and_call
    self._post_teardown()
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 1112, in _post_teardown
    self._fixture_teardown()
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 1334, in _fixture_teardown
    connections[db_name].check_constraints()
  File "C:\Users\akash\Desktop\Djangocon\django\django\db\backends\sqlite3\base.py", line 275, in check_constraints
    raise IntegrityError(
django.db.utils.IntegrityError: The row in table 'delete_dbdefaultsfk' with primary key '2' has an invalid foreign key: delete_dbdefaultsfk.language_code_id contains a value '1' that does not have a corresponding value in delete_foo.id.

======================================================================
FAIL: test_foreign_key_db_default (delete.tests.DatabaseLevelOnDeleteTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\akash\Desktop\Djangocon\django\django\test\testcases.py", line 1427, in skip_wrapper
    return test_func(*args, **kwargs)
  File "C:\Users\akash\Desktop\Djangocon\django\tests\delete\tests.py", line 899, in test_foreign_key_db_default
    with self.assertRaises(IntegrityError):
AssertionError: IntegrityError not raised
```

Reply:
Ah! I think the problem here is that we're using (a subclass of) `TestCase`, rather than `TransactionTestCase` - the `IntegrityError` doesn't get raised until the transaction ends, which is after the test method ends!

Reply:
Resolved now, added the tests here https://github.com/django/django/pull/16851/files#diff-a842e4474a8c70b274fabec1b44e377a498181dfe6c925801b7dc930917c3e03R904

Reply:
Added the test here : https://github.com/django/django/blob/d40bcc5b0b63fc1182247b8677a0d2f175e3c4da/tests/delete/tests.py#L904